### DISCUSSION TITLE
If pom.xml specifies spring-boot-starter-parent as a parent, then adding Konsist dependency makes Spring tests failing

### DISCUSSION THREAD

**Discussion by Lanjusto** [LemonAppDev/konsist#555] [Report a Problem] [answered]
### Tell us what version of Konsist you use.

0.11.0

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

I have a Maven project which specifies `org.springframework.boot:spring-boot-starter-parent` as a parent (as suggested
[here](https://docs.spring.io/spring-boot/docs/3.1.3/maven-plugin/reference/htmlsingle/#using.parent-pom)):

```xml
<parent>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-starter-parent</artifactId>
  <version>3.1.2</version>
  <relativePath/>
</parent>
```

I also have a test written Spring MVC Test framework which is green, i.e., `mvn clean package` is executed successfully.

Please see https://gitlab.com/Lanjusto/spring-with-konsist for full example.

Then I add Konsist dependency to the project to be able to write Konsist tests. To be clear, I add the dependency only, I don't add any code which uses this dependency.

### Actual behaviour

After adding Konsist dependency, I am still able to build the project and my test is still green. In other words,
`mvn clean package` is successful.

### Expected behaviour

After adding Konsist dependency, my test gets red with the following error:

`ControllerTest.simpleTest » IllegalState Failed to load ApplicationContext for [WebMergedContextConfiguration@39ad12b6 testClass = spring.with.konsist.ControllerTest, locations = [], classes = [spring.with.konsist.App], contextInitializerClasses = [], activeProfiles = [], propertySourceLocations = [], propertySourceProperties = ["org.springframework.boot.test.context.SpringBootTestContextBootstrapper=true", "server.port=0"], contextCustomizers = [[ImportsContextCustomizer@4eb45fec key = [org.springframework.boot.test.autoconfigure.web.servlet.MockMvcAutoConfiguration, org.springframework.boot.test.autoconfigure.web.servlet.MockMvcWebClientAutoConfiguration, org.springframework.boot.test.autoconfigure.web.servlet.MockMvcWebDriverAutoConfiguration, org.springframework.boot.autoconfigure.security.oauth2.client.servlet.OAuth2ClientAutoConfiguration, org.springframework.boot.autoconfigure.security.oauth2.resource.servlet.OAuth2ResourceServerAutoConfiguration, org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration, org.springframework.boot.autoconfigure.security.servlet.SecurityFilterAutoConfiguration, org.springframework.boot.autoconfigure.security.servlet.UserDetailsServiceAutoConfiguration, org.springframework.boot.test.autoconfigure.web.servlet.MockMvcSecurityConfiguration, org.springframework.boot.test.autoconfigure.web.reactive.WebTestClientAutoConfiguration]], org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@6d60fe40, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@514646ef, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.web.client.TestRestTemplateContextCustomizer@70cf32e3, org.springframework.boot.test.autoconfigure.actuate.observability.ObservabilityContextCustomizerFactory$DisableObservabilityContextCustomizer@1f, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@4b3fa0b3, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@55cb6996, org.springframework.boot.test.context.SpringBootTestAnnotation@c72b7e72], resourceBasePath = "src/main/webapp", contextLoader = org.springframework.boot.test.context.SpringBootContextLoader, parent = null]`

Please see https://gitlab.com/Lanjusto/spring-with-konsist to reproduce this behaviour.

**Comment by igorwojda** [comment]
This is unexpected. 
I am coming from Android I have to understand what's going on here. 
Would you like to help to debug this issue?

In them meantime I have started a discussion started at kotlinlang #spring channel https://kotlinlang.slack.com/archives/C0B8ZTWE4/p1695026400126699

**Comment by rlippolis** [comment]
- The `konsist` library depends on `org.jetbrains.kotlin:kotlin-compiler`, which includes a copy of the class `com.google.gson.Gson` (not sure why).
- Because of the presence of this class, Spring Boot triggers `org.springframework.boot.autoconfigure.gson.GsonAutoConfiguration`, because this config class has a `@ConditionalOnClass({Gson.class})`.
- This auto-configuration class tries to create a `GsonBuilder` bean, but fails to do so, because that class is not present on the classpath (it is present in the Gson library, but that library is not on the classpath).

This at least explains why it fails. Not sure how to solve this however. I tried adding the `com.google.code.gson:gson` library as a dependency, to see if that fixes it, but that leads to a `java.lang.NoSuchFieldError` (probably due to a version mismatch between the Gson classes in the `kotlin-compiler` dependency and the Gson library...)

**Comment by rlippolis** [comment, marked as answer]
Ah, upgrading to Kotlin 1.9.10 fixes the issue! Apparently the Gson classes were bundled accidentally in the 1.9.0 release, see: https://youtrack.jetbrains.com/issue/KT-61147

**Reply by Lanjusto** [reply]
Oh, this is a perfect solution, thanks!

**Comment by michmzr** [comment]
Its auto-configuration issue with GSON. 
You can solve it adding
`@EnableAutoConfiguration( excludeName = ["org.springframework.boot.autoconfigure.gson.GsonAutoConfiguration"])` to your class test

```
@AutoConfigureMockMvc
@EnableAutoConfiguration( excludeName = ["org.springframework.boot.autoconfigure.gson.GsonAutoConfiguration"])
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
internal class ControllerTest{

    @Autowired
    private lateinit var mockMvc: MockMvc

    @Test
    fun simpleTest() {
        mockMvc.post("/v3/accounts") {
            contentType = MediaType.APPLICATION_JSON
            content = "Hey!"
            accept = MediaType.APPLICATION_JSON
        }.andExpect {
            status { isCreated() }
        }
    }
}
```

https://stackoverflow.com/questions/50031381/why-do-i-get-gson-builder-error-when-starting-a-spring-boot-application

**Reply by Lanjusto** [reply]
Thanks, now it is clear!

**Comment by igorwojda** [comment]
So we have two fixes 🎉
I will close this one.
